### PR TITLE
generalized Wupget loading code for loading from sub folders

### PR DESCRIPTION
**Pull request by ChrisFloofyKitsune**
### Work done
Change LuaUI/barwidgets.lua and LuaRules/gadgets.lua to allow loading of their scripts from sub folders

Goals of new "subfolder" system:
* Allow needed unique assets and supporting .lua files to be in the same folder as the main Widget/Gadget (wupget) script
  * Allow for easy distribution of player created Widgets that are more than a single script file.
    * Allow widgets to be distributed as .zip (or whatever archive formats can be loaded)
    * Make eventual Steam Workshop / external mod manager integration much easier
  * Practically necessary for new UI Widgets that will be making use of RmlUi
    * each Widget will (almost always) have it's own *.rml file at the minimum
    * additionally may have *.rcss file and a few unique image files

* Have a parent-child hierarchy between loaded wupgets such that
  * if a parent wupget is disabled, all child wupgets can be easily disabled as well
  * ~~parent modules can do additional configuration of child modules~~
    * Would need a new callin somewhere or some jank work around
  * child wupgets can refer to the parent module if need be

* Do not allow a wupget to try and load other wupgets from higher up in the directory structure

a new util file ( common/springUtilities/wupgetLoader.lua ) was created to have unified code between Widget/Gadget loading so that the above goals could be met

#### Test steps
- [ ] All gadgets and widgets should happily load in as before
- [ ] the new "Subfolder Example" Widget should show up in the Widget Selector with the previously existing widgets "TEST DrawPrimitiveAtUnit GL4" and "TEST DrawPrimitiveAtUnit GL4 Minimal Example" having ">" to the left of their names to denote that they are now child widgets
- [ ] Enabling a parent widget with disabled children (red text) does **not** enable the children
- [ ] Enabling a child widget whose parent is disabled should also enable parent
- [ ] Disabling a parent widget with enabled children should put the children in a "wants to be active" state (yellow text)
- [ ] Children widgets with yellow text **should** automatically enable when the parent is enabled

### CODE REVIEW

**Comment by ChrisFloofyKitsune** [comment]
the auto removal of excess whitespace refuses to stop happening in my IDE. >_<

File: .gitignore
Comment (ChrisFloofyKitsune) [reply .gitignore:7]: hm, seems to be an artifact of me trying to get the dev setup working
now that I have it working properly- those files have not come back after I deleted them

File: common/springUtilities/scriptModuleLoader.lua
Comment (ChrisFloofyKitsune) [reply common/springUtilities/scriptModuleLoader.lua]: fixed, Echos that it's found more than one `*main.lua` file and stops the loading process there

**Comment by ChrisFloofyKitsune** [comment]
check discord thread for progress pics and commentary
https://discord.com/channels/549281623154229250/1197712298161209374

**Comment by sprunk** [comment]
consider stealing this standalone 30-liner instead https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaRules/Gadgets/api_subdir_gadgets.lua

**Comment by ChrisFloofyKitsune** [comment]
> consider stealing this standalone 30-liner instead https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaRules/Gadgets/api_subdir_gadgets.lua

doesn't have the logging I want nor cascading enable/disable for widgets as far as I can tell

anyhow, the code here is in final draft stages now

File: common/springUtilities/wupgetLoader.lua
Comment (salinecitrine) [inline common/springUtilities/wupgetLoader.lua:109]:
```suggestion
					currentChildren[#currentChildren + 1] = child
```

File: common/springUtilities/wupgetLoader.lua
Comment (salinecitrine) [inline common/springUtilities/wupgetLoader.lua:175]: Same as `tableEcho`, `loadedFilePaths` should be a set.

File: common/springUtilities/wupgetLoader.lua
Comment (salinecitrine) [inline common/springUtilities/wupgetLoader.lua:169]: With `loadedFilePaths` already checking for cycles, is checking recursion depth still necessary?

---

`recursionDepth` is also never reset, so if there's an error during loading and then loading is retried, you might end up with it starting at a non-zero `recursionDepth` value.

File: common/springUtilities/wupgetLoader.lua
Comment (salinecitrine) [inline common/springUtilities/wupgetLoader.lua:57]: I think this function can be made faster and easier to understand with a clearer tree abstraction and a depth-first search based sort; I'll try to put together some example code soon to demonstrate. I don't want to block the PR with this though, your algorithm already appears to work as expected.

File: luarules/gadgets.lua
Comment (salinecitrine) [inline luarules/gadgets.lua:361]: These are never used

**Review by salinecitrine — CHANGES_REQUESTED** [review]
Got some review done, I'll try to finish reviewing later.

Notes:
* Tested and works exactly as expected!
* Mostly just a reminder; `test_subfolder.main.lua` has been added, `test_DPAT.lua` and `test_dpat_minimal_example.lua` have been moved for testing. These changes should be reverted before merge.
* There are some mostly unrelated formatting and comment changes that should at least be separated out into their own commits (I think this is done for some but not all of these changes).

File: common/springUtilities/wupgetLoader.lua
Comment (ChrisFloofyKitsune) [reply common/springUtilities/wupgetLoader.lua:169]: I think the idea a couple months ago was... to yell at people nesting things that deeply

but since cycles are being protected against someone would need to intentionally make this run forever
I'll take it out

File: common/springUtilities/wupgetLoader.lua
Comment (ChrisFloofyKitsune) [reply common/springUtilities/wupgetLoader.lua:57]: eh, I made this up on the spot when getting things to sort how I wanted in the Widget Selector
then I reused it here

Comment (ChrisFloofyKitsune) [reply common/springUtilities/wupgetLoader.lua:57]: I don't know what a depth-first search based sort would look like here
I'm just trying to lean on the built-in sort as much as possible

Comment (salinecitrine) [reply common/springUtilities/wupgetLoader.lua:57]: https://github.com/salinecitrine/Beyond-All-Reason/commit/25964c233be5fc6a18b03b064acfba4d68d435d6

This is what I had in mind as an alternative sorting algorithm. I did some benchmarks, and it's about 17% slower for mostly flat widget trees (as it is now), and about twice as fast for trees with a lot of children. Not a requirement, mostly just wanted to try it out.

---

I realized something while writing it out, though: isn't there a conflict between loading widgets in parent/child order, and layer order? what if a widget tree has one widget that wants to be layer=-100, won't it be restricted to loading after its parent?

Comment (ChrisFloofyKitsune) [reply common/springUtilities/wupgetLoader.lua:57]: You did performance testing? You are a more diligent developer than I.

I doubt that widget trees will be getting very deep. But who knows what the future holds.

-------

Then the parent should have a lower layer too? That or it shouldn't be a child widget?

hmm...

I now see a case for allowing multiple  "*.main.lua" files in a folder- so they can be organized together but not be forced to have a parent/child relationship

I should also write in more documentation in the example widget, at least putting in a comment that shows returning a list of paths

Comment (ChrisFloofyKitsune) [reply common/springUtilities/wupgetLoader.lua:57]: I don't want to go into the complexity of dependency resolution stuff if it can be helped

File: luaui/barwidgets.lua
Comment (salinecitrine) [inline luaui/barwidgets.lua:521]: Is there a case where it's not already active?

File: luaui/barwidgets.lua
Comment (salinecitrine) [inline luaui/barwidgets.lua:940]: Should this enable-parent-section go in the `if not ki.active then` block below? I'm not actually sure which; I guess it shouldn't be possible to have the child enabled but not the parent, so it might not matter?

File: luaui/barwidgets.lua
Comment (salinecitrine) [inline luaui/barwidgets.lua:939]: remove commented code

File: luaui/barwidgets.lua
Comment (salinecitrine) [inline luaui/barwidgets.lua:980]: probably remove the tableEcho.

I did actually trigger this somehow (the screenshot I added was the initial state), but now I can't get back to that state :(

**Review by salinecitrine — COMMENTED** [review]
* Finished reading through the code, and added a few more comments.
* I'm a little nervous about changing all the widget/gadget loading (particularly the parent/child stuff), so maybe a few other people could test this as well, just to make sure it works for them.

File: luaui/barwidgets.lua
Comment (ChrisFloofyKitsune) [reply luaui/barwidgets.lua:521]: this is one of the many leftovers from the previous code
the active field on the widgetInfo is not set until that line

File: luaui/barwidgets.lua
Comment (ChrisFloofyKitsune) [reply luaui/barwidgets.lua:940]: it is logically inconsistent/weird to have it outside that code block
I'll move it

File: luaui/barwidgets.lua
Comment (salinecitrine) [inline luaui/barwidgets.lua:814]: Oops, I actually meant to comment on this `ki.active = true` (instead of https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2546#discussion_r1518118380).

---

Is there a case where it's not already active?

File: luaui/barwidgets.lua
Comment (salinecitrine) [reply luaui/barwidgets.lua:526]: Oh, oops, I actually meant to leave this comment on another line. Added a comment there.

**Review by salinecitrine — COMMENTED** [review]
Changes look good to me!

The thing I'm most uncertain about is probably the sorting in `wupgetLoader.lua`; it would be nice to get more input on how layer and parent/child should factor into loading order.

One thing I forgot to mention is that I noticed that you can play around with `fontSpace` in `widget_selector.lua` to get the branches to connect better (I couldn't get it perfect without a monospace font, but your FSP thing does help a lot)

File: luaui/barwidgets.lua
Comment (ChrisFloofyKitsune) [reply luaui/barwidgets.lua:814]: it does not look like there is ever a case where `active == false` by that point

**Comment by ChrisFloofyKitsune** [comment]
> The thing I'm most uncertain about is probably the sorting in `wupgetLoader.lua`; it would be nice to get more input on how layer and parent/child should factor into loading order.

with these changes, the idea is that parent widgets will load first so that it can be used to contain some state for the children widgets

looking at the previously existing code though- it does seem that the update order of widgets will go entirely off of layer order

**Comment by ChrisFloofyKitsune** [comment]
Added testing for and confirmed that widgets still use layer order for Callins
image

**Comment by ChrisFloofyKitsune** [comment]
let me know when to delete the loader_testing folder
or a contributor can just do it themselves before merging

**Comment by ChrisFloofyKitsune** [comment]
Closing this as it's gotten quite old and I would want to go over everything again anyhow.

**Comment by WatchTheFort** [comment]
Has anything changed to make the code obsolete? Is this a requirement for rmlui integration? I recall it being a requirement for proper modding support.

**Comment by ChrisFloofyKitsune** [comment]
There's been a fair amount of extra code added for debug purposes and I wrote this months ago when I was new to the project. I know more about it now and I've half forgotten what went into this PR anyhow.

Eventually, something like this is gonna need to exist for proper modding support for sure though. The main initial goal was to try and organize the 100+ widgets so that they're easier to find and work on. But then I was told that "just use the IDE to search in the folder according to prefix" and it works well enough that my initial driving reason for making this died out.

Also there was a few edge cases I never got sorted out like- what to do if a user widget is added with the same name as something that's already been added to the hierarchy tree? I would need to rewrite a lot of stuff anyhow- doing things like keeping track of which archive things came from and doing weird VFS stuff.

It is not a hard requirement for RmlUI integration- but would definitely make things neater.

**Comment by sprunk** [comment]
I'll repeat my suggestion to steal this standalone 30-liner instead https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaRules/Gadgets/api_subdir_gadgets.lua - reads `main.lua` from each subfolder. Later on you can expand upon it and add features as needed.

**Pull request closed without merging**
